src/boat_simulator/tests/unit/nodes/low_level_control/test_controller.py

```diff
@@ -0,0 +1,185 @@
+        (50.488, -36.78, 30, 0.5, 0.7, 0.34, 1),
+    ]
+)
+def rudder_controller(request):
+    # Initialize the RudderController object with parameters from the fixture
+    current_heading, desired_heading, current_control_ang, time_step, kp, cp, control_speed = (
+        request.param
+    )
+    return RudderController(
+        current_heading=current_heading,
+        desired_heading=desired_heading,
+        current_control_ang=current_control_ang,
+        time_step=time_step,
+        kp=kp,
+        cp=cp,
+        control_speed=control_speed,
+    )
+
+
+def test_compute_error(rudder_controller):
+    # Test compute_error method of RudderController
+    error = rudder_controller.compute_error()
+    desired_rad = np.deg2rad(rudder_controller.desired_heading)
+    current_rad = np.deg2rad(rudder_controller.current_heading)
+    exp_error = atan2(sin(desired_rad - current_rad), cos(desired_rad - current_rad))
+    assert np.equal(error, exp_error)
+
+
+def test_compute_setpoint(rudder_controller):
+    # Test compute_setpoint method of RudderController
+    setpoint = rudder_controller.compute_setpoint()
+    heading_error = rudder_controller.compute_error()  # heading_error in radians
+
+    rudder_setpoint = (rudder_controller.kp * heading_error) / (
+        1 + (rudder_controller.cp * abs(heading_error))
+    )
+
+    if abs(rudder_setpoint) > RUDDER_MAX_ANGLE_RANGE[1]:
+        raise ValueError("rudder_change must be between -pi/4 and pi/4")
+
+    rudder_setpoint_deg = np.rad2deg(rudder_setpoint)  # in degrees
+    running_error = rudder_setpoint_deg - rudder_controller.current_control_ang  # in degrees
+    assert np.equal(setpoint, rudder_setpoint_deg)
+    assert np.equal(running_error, rudder_controller.running_error)
+
+
+def test_reset_setpoint(rudder_controller):
+    # Test reset_setpoint method of RudderController
+    new_desired_heading = 60  # Example value
+    expected = rudder_controller.reset_setpoint(new_desired_heading)
+    rudder_controller.change_desired_heading(new_desired_heading)
+    calculated = rudder_controller.compute_setpoint()
+    np.equal(calculated, expected)
+
+
+# testing for one iteration only
+def test_update_state(rudder_controller):
+    rudder_controller.compute_setpoint()
+    assert rudder_controller.update_state() == (abs(rudder_controller.running_error) == 0)
+
+
+def test_update_state_continuous(rudder_controller):
+    rudder_controller.compute_setpoint()
+    progress = False
+    counter = 0
+    if np.isclose(rudder_controller.running_error, 0, 0.1):
+        progress = True
+    else:
+        while not np.isclose(rudder_controller.running_error, 0, 0.1):
+            counter += 1
+            rudder_controller.update_state()
+            if rudder_controller.running_error == 0:
+                progress = True
+                break
+            if counter > 1000:
+                break
+    assert progress
+
+
+def test_update_state_reset(rudder_controller):
```

Comment:
You should also test what the rudder angle should be while it actuates in addition to the running error